### ISSUE TITLE
Bring back ability to `install` and `get` on arbitrary path

### ISSUE THREAD

**Issue by yarikoptic** [datalad/datalad#3759]
#### What is the problem?

Originally introduced in #3746 and built on top in #3747 was RFing to remove use of AnnotatePaths.  It resulted in removing a possibility to operate on paths to be `install`ed or `get`ed, when CWD is outside of their (single) super dataset.

I don't think it is not productive for me to stop progress in those PRs by demanding retaining backward feature parity.  That is why I decided to move discussion into this separate issue.

# Use cases which were made impossible with those changes

All boil down to the original promise that, unlike git/git-annex, datalad could operate seamlessly on a path while residing outside of the dataset itself.  It provides a convenience for scripting etc.  Introduced changes make following use cases, which I used from time to time, not possible

* ```datalad -f '{path}' search -d /// haxby | xargs datalad install```
would need to become e.g. 
   - ```cd ~/datalad && datalad -f '{path}' search haxby | xargs datalad install```
* ```datalad install ///openneuro/ds000001 && datalad get ds000001/sub-01/anat```
would need to become e.g. 
  - ```datalad install ///openneuro/ds000001 && cd ds000001 && datalad get sub-01/anat```
  - ```datalad install ///openneuro/ds000001 && datalad -C ds000001 get sub-01/anat```
* pretty much for any "let me just get that file in my other collection of datasets" I would not be able to simply provide a path to it while keeping working in my project. With our new treatment of relative paths I would likely need to duplicate the path or really `cd` there (could be via `-C` through) to perform that operation
* ```datalad install ///openneuro/ds00000{1,2,3} && datalad get ds00000*/sub-01/anat```
would need to
  - establish top level super dataset first
    - ```datalad create throwaway && cd throwaway && datalad install -d . ///openneuro/ds00000{1,2,3} && datalad get ds00000*/sub-01/anat```
    - or some ugly looping

Looking at the diff I think it should be feasible to bring back some part of the annotate_paths notion (just to deduce a set of top level superdatasets given a list of paths), and thus bring back prior functionality.  I am yet to decide either it is really needed or we could get away with it.

If we are to bring it back (in a new implementation), here is a note that a test verifying this feature was removed and would be nice to bring it back: https://github.com/datalad/datalad/pull/3747/files#diff-8e9c3c169ed7a22baba947b11ad9f402L400

**Comment by mih** [comment]
Just to clarify what is the underlying change that *could* have the side-effects outlined above: The new code no longer supports operating on paths that belong to disjoint superdatasets. Out of all use cases listed above, only the last one kinda falls into this category. Not really, because it still operates on the `///` superdataset only, despite leaving three separate resulting datasets around.

```
datalad install ///openneuro/ds000001 && cd ds000001 && datalad get sub-01/anat
```
is a more complicated variant of

```
datalad install ///openneuro/ds000001 && datalad get -d ds000001 ds000001/sub-01/anat
```

which closely mimics the Python API.

As expressed elsewhere, I don't mind the `install` command being an outlier and doing something more fancy than other commands. Even now, it neither uses the `annotate_paths()` helper, nor the new approach.

IMHO the `install` API and feature set should get a close look, e.g. in the light of feature requests like https://github.com/datalad/datalad/issues/2109 The combinatorics of all plans are a bit overwhelming for me, and I do not have a clear picture of what needs to be implemented and supported. However, I would much prefer, if the implied complexities do not transcend beyond `install()`.

**Comment by yarikoptic** [comment]
> The new code no longer supports operating on paths that belong to disjoint superdatasets.

not only -- it doesn't support operating from a directory which is outside of the dataset(s) to which provided paths belong to. And that is where all initial use cases fall under.

Re:
> ```
> datalad install ///openneuro/ds000001 && datalad get -d ds000001 ds000001/sub-01/anat
> ```
> 
> 
> which closely mimics the Python API.

not entirely.  In Python API, which would be
```python
dl.install('///openneuro/ds000001').get('sub-01/anat')
```
there would be no need for duplication (of `ds000001`) or explicit `cd`, since paths will be relative to the dataset dir, so no duplication would be needed altogether, not even twice (like in current cmdline interface).  The only way would be is to really `cd` (or -C) into that dataset to avoid 3rd duplication.

**Comment by mih** [comment]
Yeah, but that is not the key issue. We can easily enhance `require_dataset()` to scan all the `path` args to determine the closest superdataset they all fit it. That would be fine, as it does not impose any additional complexity on the command implementation. It still only deals with a single reference dataset. It could be slow (as before), but it would be just a mode, not the mode.

**Comment by yarikoptic** [comment]
yeah, I think this should be the way to go.  It would probably address >90% of real use cases.  For the other 10% we could always praise YODA which would require having the superdataset to do things "right" ;)

**Comment by mih** [comment]
I looked into a possible implementation of a `require_dataset()` that determines the intended superdataset from inputs paths, and I no longer think that this would be a good idea.

ATM we consider the `dataset` arg, and if that wasn't given, we check PWD. If we introduce `paths` in the mix, we need to decide when to consider them, and how to resolve conflicts?

What if I call

`datalad save /existselsewhere/p1`

Does that mean I intend to save `p1` in the dataset it exists in? Or is that an error, because there is a Git repo at PWD?

What if I call

`datalad save /super/sub/p1 /super/sub/subsub/p2`

Does that mean I want everything saved all the way up to `/super`, or just to `/super/sub`, or just the paths in the respective subdatasets? What if PWD at the time of execution is `/super/sub`, what if it is `/super/sub/subsub`?

I think we would end up in a situation (again), where it is very hard to issue actual error messages, and prevent unintentional behavior, because everything is so flexible that it becomes hard to decide if something was somehow intentional, despite being unlikely.

I think we should rather alter `install`, and only `install` to make some of the usecases above possible. Going over your list:

* ```datalad -f '{path}' search -d /// haxby | xargs datalad install```

   We introduce an internal loop in `install` that processes each input arg independent of any other, and finds the dataset that single path is in, if no `dataset` arg is given.

* ```datalad install ///openneuro/ds000001 && datalad get ds000001/sub-01/anat```

   with the change above it becomes:

  `datalad install ///openneuro/ds000001 && datalad install -g ds000001/sub-01/anat`

   although THB, I consider the present solution just as good, if not clearer:

   `datalad install ///openneuro/ds000001 && datalad -C ds000001 get sub-01/anat`

* ```datalad install ///openneuro/ds00000{1,2,3} && datalad get ds00000*/sub-01/anat```

   also solved via the internal `install` arg-loop, when `get` is replaced by `install -g`

In conclusion, the new setup has removed a large range of argument ambiguity issues and substantially simplified the complexity of processing. That makes me very reluctant to go back on a global scale.

As I expressed before, I have no issues in tuning `install` to arbitrary degrees. We have `clone` that is more Git-like in behavior, but still understands all the semantics of datalad-specific arguments and processing. Moreover, `install` is essentially not used internally. That gives us the freedom to alter and enhance `install` substantially.

**Comment by bpoldrack** [comment]
> I looked into a possible implementation of a require_dataset() that determines the intended superdataset from inputs paths, and I no longer think that this would be a good idea.

Yes, I came to the same conclusion in PR #3846.

Generally just want to point out, that part of the discussion  and implied issues is to be found the aforementioned PR.

Apart from that: I agree. The original approach that the issue wanted to "bring back" just leads to a lot of ambiguous situations. Any solution to that will likely make sense to whoever wrote it only.

**Comment by yarikoptic** [comment]
> ... and how to resolve conflicts?

not sure what conflicts you have in mind, may be you had "mental" conflicts due to the stated by Ben

> ... leads to a lot of ambiguous situations

Things are "ambiguous" IMHO only if there is no clear, used by **all** commands, description of rules on how dataset and path(s) are treated, and what to expect from `datalad`. 

NB let's first postpone discussion of possible implementation difficulties and first agree on the description of desired rules/behavior, even though it might smell a bit too much of "good old" AnnotatePaths.

Why it could not be some list of rules like:

1. If no working dataset is explicitly specified (via `dataset` argument), a dataset current directory belongs to is taken as the "working dataset".  (note: If there is no "working dataset", that is ok (rules `4., 5.` below have "if"))
2. Any given command first operates on the path(s) within the dataset where path(s) belong to (clarification: a (sub)dataset path causes operation within that (sub)dataset; even `datalad install .` is ok).
3. `--recursive` flag could be used (where applicable) to make command operate within subdatasets of the provided path(s). 
4. If there is a "working dataset", changes on the paths performed in the sub-datasets of the "working dataset" are also saved in any sub-dataset on the path from the "working dataset" to the dataset specified path(s) belongs to (old behavior I think)
5. For the paths outside of the "working dataset" (if any), changes to their datasets are not saved in their possible super-datasets (old behavior)
6. Merely a note: some commands (e.g. `get` and `install`) might require necessary sub-datasets first being `install`ed to act on the path(s) specified on the command line.  Such sub-datasets will be installed regardless of relation to the working directory dataset (old bevavior, nothing really special here). `create` might require to create a directory in the specified path to act on.

I think above rules make it clear how my and @mih's examples should work.  Or am I missing something?

## Possible additional "rule"

I thought to add

1a. If no path is provided, operation on "working dataset" path is performed

That could explain current behavior of plain `datalad save`, `status`, and `diff` but I think it might cause undesired consequences to `get`, `install` and other commands while people would use datalad programmatically and arrive with an empty list of arguments to operate on.  Although, from the other hand, I do not see why/how `save` et al should be special - kinda the same programmatic gotcha.  And that is where probably if we decide to stick to current behavior where command-specific behaviors could be described.

1a. If no PATH argument is provided, commands `save`, `status`, `diff` operate on the path of the "working dataset".

Behavior of other commands is actually now a bit more convoluted:

1b. If not PATH argument is provided, commands `get`, `install`, ... operate on the path of the "working dataset" only if it was explicitly specified via `--dataset` argument, and fail to do anything otherwise.

So, `1.` is not yet universally in place, because IIRC we still behave differently depending on either explicit `-d` was specified or not. E.g. `get` would fail to do anything without paths while I am already in a dataset. We should either unify or clarify

## Disabled use-case(s):

`2.` is kinda the tricky one and where @mih tried the paths tricks (`/.`) to disambiguate between "within the dataset" or "path of the dataset within its super-dataset if I there right now".  I am no longer sure, if with all the "save it all" and introduction of `1.` we really need to distinguish those two cases. The only usecase I currently see (correct me if I am wrong) which would be made unavailable then would be 

- save changed state of the sub-dataset without saving (new/other) changes 

If there is no other corner case like that it could be mitigated by either advising to use git directly in such cases or adding a config or some `save` option (like `--disable-rule-2` which would in effect achieve desired action by the rule `4.`, or `--act-on-subdataset-paths` or alike) if really desired to consider paths pointing to subdatasets as just paths within their super-datasets.

**Comment by mih** [comment]
I have tried a few times to load the problem into my head, and failed. I think I am unable to answer the question "Why it could not be some list of rules like..." without doing a full inspection of all commands with respect to compliance and compatibility with these rules. Without that knowledge, it is hard to *agree* on things, as it is more or less pointless if we agree, but it still doesn't work, or rather, we cannot find anyone that has the mental capacity and available time to actually do it. This is pretty much the situation we are in. We have three generations of API, no transitions completed, and we have stopped progressing because no ultimate solution can be found implemented.

If we want to go further down this path, we have to be ultra-specific to make progress. For example:

> --recursive flag could be used (where applicable) to make command operate within subdatasets of the provided path(s).

Does that mean a `datalad status subds1/subds2/subdir` would need `--recursive`? If not, what needs `--recursive` exactly. If so, how to then say "only the status within `subdir/`, but not within further subdatasets underneath them?

I have the feeling that keeping to find the cases that **don't** work now, and fix them, rather then going for 4th API generation in addition is a more fruitful approach. All the status-based commands that we have now are fairly consistent from my POV, and the original cases reported in this issue do not seem to outline general issues with the concept AFAICT.

Given that this issue is also the last blocker for 0.12, we are also talking about the release date here: in days, weeks, months, or longer. The work towards the new core commands has started more than a year ago (that is the lifetime of the 0.9 and 0.10 series combined). We merged that work in ~May 2019 into master. We already had a 6-months consolidation period. Now we seem to be talking about starting over.

**Comment by yarikoptic** [comment]
Will reply "fuller" later, burning now, but re

> > --recursive flag could be used (where applicable) to make command operate within subdatasets of the provided path(s).
> 
> Does that mean a `datalad status subds1/subds2/subdir` would need `--recursive`? If not, what needs `--recursive` exactly. If so, how to then say "only the status within `subdir/`, but not within further subdatasets underneath them?

isn't that exactly the case here for 2. + 3. rule:

2. go to `subds1/subds2`, run status on `subdir`
3. if you want status to include subdatasets under `subdir` - add `--recursive`

?

> I have the feeling that keeping to find the cases that **don't** work now, and fix them

fix them to follow which rules?

**Comment by mih** [comment]
Ok, now I have the impression that I got this wrong. We are talking about *describing* how the present commands work, not about *defining* how they should work, correct? I am all for that.

But we should rather not do this here then, because this issue is the last blocker for the 0.12 release.

**Comment by yarikoptic** [comment]
> Ok, now I have the impression that I got this wrong. We are talking about _describing_ how the present commands work, not about _defining_ how they should work, correct? I am all for that.

I am no longer sure how current command works since there is no clear description anywhere.  And there could be no clear description because different commands work differently. I know that some rules are not really working (hence this particular issue was filed).  In the https://github.com/datalad/datalad/issues/3759#issuecomment-562595209 comment I just answered your question in terms of rules.

I see this issue as a blocker since it really forbids previous common (in my world) behavior which was broken only recently.

**Comment by mih** [comment]
We never had a description so that is not the breakage. What about the specific action that I proposed to reintroduce the abilities to install that I proposed?

**Comment by yarikoptic** [comment]
> We never had a description so that is not the breakage.

when something was working before (probably from version 0.3 or so) and now doesn't -- that is a definition of "breakage" in my life. Indeed though we had no formalized description of expected behavior (which is what I am complaining here about as well), but e.g. we had http://docs.datalad.org/en/latest/basics.html:  "running `datalad install -d/// crcns` will install the crcns subdataset under $HOME/datalad/crcns. This is the same as running `datalad install $HOME/datalad/crcns`." which is generally no longer the case.

I think the main reason for such original behavior was from the original goal for datalad -- to provide "data distribution". It is analogous to software distributions, where I do not need to `cd` to the package content directory to install/uninstall it. So we had it implemented so it would be possible to install/uninstall/get/drop/... distribution components, and it was conscious design decision, even though we seems have not clearly documented it.

> What about the specific action that I proposed to reintroduce the abilities to install that I proposed?

Are you talking about "I think we should rather alter `install`, and only install to make some of the usecases above possible" from https://github.com/datalad/datalad/issues/3759#issuecomment-559759583?  That one doesn't address anything for `get` which is a frequent use case for me.

## Alternative 

Could be to (at least for 0.12) to revert #3746 and #3747 and reopen #3551 (was there anything else?) and fix it differently?

> In conclusion, the new setup has removed a large range of argument ambiguity issues and substantially simplified the complexity of processing.

I do not see any example of current (as of 0.11.x or before the breakage) "ambiguous" behavior of `get` or even `install`!   Do not mix unambigous with inconvenient (e.g. installing within subdirectory while needing to specify the top of the dataset or full path).

I do appreciate that re-implementation might have indeed "substantially simplified" internal implementation, but let's not pretend that the later has done something to achieve the former in user perspective. If anything, it made it more confusing to me (a user) now: if before I could just request to `get` a file, now I need first to disambiguate -- am I in the (super)dataset that the file "belongs" to?  if I don't know where the boundary of some (super)dataset on the path is I would need to first cd to the directory to run `get`. If I don't know if multiple paths are reachable from some super-dataset, I would need first to go to some common to all of them directory.  That is what I would call "ambiguous", not the previous behavior!

> I looked into a possible implementation of a `require_dataset()` that determines the intended superdataset from inputs paths, and I no longer think that this would be a good idea.

Again, this sounds like a bottom-up verdict, instructed by technicalities, on user-visible behavior without first making a decision on desired behavior/limitations it introduces.  

## My verdict:

Overall, I give up on trying to retain consistency of 0.12 with (unambiguous) previous behaviors in this case. I do not want to stop 0.12, and I can adapt (although probably with some cursing for a variety of usecases since proposed workarounds at user level are IMHO ugly (break completion etc) even if you say they are "consistent"). 
You are welcome to remove this issue from the milestone/0.12 blockers.  Just keep it open until it is fixed (or let's say a year timeout if nobody else cares) and add a loud CHANGELOG statement about change in behavior  such as "Dear users, we are now even closer to how Git operates -- you must cd or specify dataset first .... (see #3759 for discussion)". Also please re-check with openneuro, if it breaks nothing in their case (but given number of issues I keep unraveling for openneuro datasets, I am not quite certain on how much integration testing they have).
  
Meanwhile I will copy proposal about rules into a separate issue. I do strongly believe that we need to agree at that principle level first and then see how it could be implemented, not the other way around (implement some handling and see what it brings us).  I do have a feeling that we would arrive at some mid-point between `AnnotatePaths` and `require_dataset`

Cheers

**Comment by mih** [comment]
> Are you talking about "I think we should rather alter `install`, and only install to make some of the usecases above possible" from [#3759 (comment)](https://github.com/datalad/datalad/issues/3759#issuecomment-559759583)? That one doesn't address anything for `get` which is a frequent use case for me.

Yes, this is what I was talking about.

> ## Alternative
> Could be to (at least for 0.12) to revert #3746 and #3747 and reopen #3551 (was there anything else?) and fix it differently?

I am open to any motion. I personally don't see how to achieve an alternative fix without a substantial investment of energy, but that surely doesn't imply that it isn't possible, or wouldn't be better.

> > In conclusion, the new setup has removed a large range of argument ambiguity issues and substantially simplified the complexity of processing.
> 
> I do not see any example of current (before the breakage) "ambiguous" behavior of `get` or even `install`! Do not mix unambigous with inconvenient (e.g. installing within subdirectory while needing to specify the top of the dataset or full path).

I was specifically talking about the implementation, not the user visible surface. I felt that the only person that could understand `AnnotatePaths` was me, and I can confidently say that I do not understand it. So even if it "works" I don't know why and I consider it unfixable from my POV. I felt that bug reports were accumulating, and nobody (including me) could or would do anything about it. That, and the fact that I desire to remove incomplete implementations, were the motivation to RF `get/install`. If that was an error of judgement, I apologize. Feel free to revert as much as you see fit.

> Overall, I give up on trying to retain consistency of 0.12 with (unambiguous) previous behaviors in this case. I do not want to stop 0.12, and I can adapt (although probably with some cursing for a variety of usecases since proposed workarounds at user level are IMHO ugly (break completion etc) even if you say they are "consistent").
> You are welcome to remove this issue from the milestone/0.12 blockers. Just keep it open until it is fixed (or let's say a year timeout if nobody else cares) and add a loud CHANGELOG statement about change in behavior such as "Dear users, we are now even closer to how Git operates -- you must cd or specify dataset first .... (see #3759 for discussion)". Also please re-check with openneuro, if it breaks nothing in their case (but given number of issues I keep unraveling for openneuro datasets, I am not quite certain on how much integration testing they have).

I (continue to) have no desire to force change. I am just out of ideas how to address all expressed desires simultaneously.

**Comment by mih** [comment]
Partial fix attempted in #3918

**Comment by oesteban** [comment]
I believe this is a problem related to this issue:

```

> /root/capsule/code/python-client/templateflow/api.py(203)_datalad_get()
-> api.get(path=str(relpath), dataset=str(TF_LAYOUT.root))
(Pdb) str(relpath)
'tpl-MNI152Lin/tpl-MNI152Lin_res-01_desc-brain_mask.nii.gz'
(Pdb) str(TF_LAYOUT.root)
'/templateflow'
(Pdb) n
datalad.support.exceptions.IncompleteResultsError: Command did not complete successfully. 1 failed:
[{'action': 'get',
  'message': ('path not associated with dataset %s', Dataset('/templateflow')),
  'path': '/root/capsule/code/python-client/tpl-MNI152Lin/tpl-MNI152Lin_res-01_desc-brain_mask.nii.gz',
  'status': 'error'}]
```

In other words:
```
>>> datalad.api.get(
...     path='tpl-MNI152Lin/tpl-MNI152Lin_res-01_desc-brain_mask.nii.gz', 
...     dataset='/templateflow',
... )
datalad.support.exceptions.IncompleteResultsError: Command did not complete successfully. 1 failed:
[{'action': 'get',
  'message': ('path not associated with dataset %s', Dataset('/templateflow')),
  'path': '/root/capsule/code/python-client/tpl-MNI152Lin/tpl-MNI152Lin_res-01_desc-brain_mask.nii.gz',
  'status': 'error'}]
```
when the cwd is `/root/capsule/code/plython-client`. I doubt this is intended.

If you `os.chdir("/templateflow")` there is no issue anymore. Seems like the `dataset` argument is actually dismissed?

**Comment by yarikoptic** [comment]
Not really related - it is just a matter of datalad API sophistication (and may be insufficient in prominence documentation on this aspect): if dataset argument is a string, not a dataset instance, relative paths are relative to cwd. So I would just do... (need to get to laptop)
